File: angles_only.m
% Returns the azimuth and elevation in radians [2x1] Array
% R = diag([(0.5 * pi / 180)^2, (0.5 * pi / 180)^2]);
% add_noise is either {true, false}
function measurements = angles_only(positions, mu, R, add_noise)
    x = positions(:,1) + mu;
    y = positions(:,2);
    z = positions(:,3);
    
    alpha = atan2(y, x);
    epsilon = asin(z ./ sqrt(x.^2 + y.^2 + z.^2));
    measurements = [alpha epsilon];
    
    if add_noise 
        measurements = mvnrnd(measurements, R);
    end
end

File: xradar_uq/measurement_systems/angles_only.py
import equinox as eqx
import jax
import jax.numpy as jnp
from beartype import beartype as typechecker
from jaxtyping import Array, Bool, Float, Key, jaxtyped
from xradar_uq.measurement_systems import AbstractMeasurementSystem


class AnglesOnly(AbstractMeasurementSystem, strict=True):
    """
    This measurement system takes in an array whose first three indices are the spatial dimensions.
    Then returns the distance, elevation and azimuth.
    """
    
    mu: float = 0.012150584269940
    covariance: Float[Array, "2"] = eqx.field(
        default_factory=lambda: jnp.diag(jnp.array([(0.1 * jnp.pi / 180) ** 2,
                                                    (0.1 * jnp.pi / 180) ** 2]))
    )


    @jaxtyped(typechecker=typechecker)
    @eqx.filter_jit
    def __call__(
        self,
        state: Float[Array, "state_dim"],
        key: Key[Array, ""] | None = None,
    ) -> Float[Array, "2"]:
        
        # Satellite position in barycentric coordinates
        satellite_pos = state[:3]
        
        # Earth position in barycentric coordinates
        earth_pos = jnp.array([-self.mu, 0.0, 0.0])
        
        # Satellite position relative to Earth
        relative_pos = satellite_pos - earth_pos
        x, y, z = relative_pos[0], relative_pos[1], relative_pos[2]
        
        # Range, azimuth, elevation from Earth
        alpha = jnp.arctan2(y, x)          # Azimuth
        epsilon = jnp.arcsin(z / jnp.sqrt(x**2 + y**2 + z**2))      # Elevation

        measurements = jnp.array([alpha, epsilon])
        if key is not None:
            measurements = jax.random.multivariate_normal(
                key, mean=measurements, cov=self.covariance
            )

        return measurements
